### PR TITLE
FIX:updates due to midi read/write port now being char*, optional ALSA mode

### PR DESCRIPTION
**Pull request by TheTechnobear**

### CODE REVIEW

File: core/Midi.cpp [hunk 1]
```diff
@@ -124,7 +137,13 @@ void Midi::midiOutputLoop(){
 void Midi::readInputLoop(){
 	while(!gShouldStop){
 		int maxBytesToRead = inputBytes.size() - inputBytesWritePointer;
-		int ret = read(inputPort, &inputBytes[inputBytesWritePointer], sizeof(midi_byte_t)*maxBytesToRead);
+        int ret = -1;
```
Comment (giuliomoro) [inline core/Midi.cpp]: @TheTechnobear everything looks good, thanks. Let's try to stick to tabs for indentation next time!

**Comment by TheTechnobear** [comment]
@giuliomoro  ouch, thats a royal pain... all projects i work on use spaces, i thought that was pretty much the 'defacto standard' these days....
what would be handy, is if you set up a code formatter... such that we can just run the code through before submission. (eclipse supports this, but there are other command line variants).
I think otherwise its pretty hard to keep it clean in open source, as a contributor, every projects has its own style :(

**Comment by giuliomoro** [comment]
To transform four spaces into one tab you can add this to a pre-commit hook
`vim -c 'set noet|retab!|wq' theFileName`

**Comment by TheTechnobear** [comment]
yeah, I'm a bit wary of transformations on pre-commits...  
(I had 'code formatter' on pre-commit  go wrong in my early career, which wiped out a days worth of work!) 
Ive been using 'astyle' for another project, which might work nicely
(http://astyle.sourceforge.net, free/open source, and cross-platform)

I think your style is pretty simple, so I don't think it would be hard to come up with the correct 'options' for formatting...

http://astyle.sourceforge.net/astyle.html, then its just a simple matter of running this before commits. would be very simple the for contributors to do.

the only 'downside' is initially, there may be a few commits that are just code formatting.
(given don't actually have much code, it might be worth running this against all files once, and then committing this as a single commit, to get it 'out of the way')
of course with the right options, and assuming there has been consistency in the past, then there wont be that many changes.


(the advantage astyle over using eclipse is its editor agnostic)

**Comment by giuliomoro** [comment]
@TheTechnobear I got around testing the ALSA API, while also fixing a stupid error in the code (writeOutput, which is supposed to be real-time safe was actually writing to the hardware device, 	bf34303).

I see that the CPU usage behaves strangely for ALSA. (as the read/write tasks run in kernel mode (secondary mode), I refer here to CPU usage reported by top (``top -p `pgrep projectName` ``) ((the audio-thread CPU usage is about 24% in both cases))
* For non-ALSA, CPU usage is stable around 6-7%. 
* for ALSA, CPU starts around 12% and then keeps increasing as more inputs are received over time, e.g.: it seems that the CPU usage is somehow proportional to the number of past events (not to the number of events in the latest few milliseconds).

Did you notice this / can you have a look at this?

**Comment by TheTechnobear** [comment]
just tested this... 

No, I don't see this behaviour at all using alsa.

what I did was use communication, basic-midi*  and turn it to alsa mode.
I see around 3.3-3.7% with no midi, and then it goes to around 4% .

I tried 2 different devices, my Virus TI and Ableton Push 2 and both behaved the same.

I also tried the Push 2, with the non-alsa implementation, and got exactly the same cpu load.

so im not sure why your seeing this.... my only thoughts are:
-  Ive only run this against the code i issued as a PR, not the current dev build which may have other changes. similarly ive not updated bela.
- it could be device specific? (but this seems unlikely given the raw file works ok)



* sorry, I dont really have a better test case, as i don't use PD, and there are no polyphonic/real synth examples... and Ive not had time to dig into this yet.

**Comment by TheTechnobear** [comment]
Ok, so Ive done a pull of the dev build... and things are generally much worst on that :( 
(its something else not alsa)

so the same above test, what I notice is:

a) without ALSA, the process starts at 6% (so almost double previous tests), then jumps to 12-13% as soon as midi comes in.

b) with ALSA, starts at 12%, and indeed rises ... but when you stop playing notes it drops.

as far as I can see there is a more general regression in the audio thread handling... as the basic test has gone from 4% to 12% , even for non-alsa, clearly something is not right.
I suspect if you find the cause of that, then alsa will drop too... since prior to these other changes the  alsa implementation pretty much performed identically to the non-alsa.

**Comment by TheTechnobear** [comment]
ok, so be thorough, I rolled back to the commit where you did the merge, which is
[67e7eb7caecb0117677334e67e3c3214ff76fed8](https://github.com/BelaPlatform/Bela/tree/67e7eb7caecb0117677334e67e3c3214ff76fed8)

and re-run the tests, and can confirm my initial findings...

what is also striking is now with Bela you are at 6% cpu load before any note is played, so before an oscillator is even sounding, at the previous point this was 0%.... so something is taking 6% cpu with no alsa, and 12% with alsa.  

Ive not been following what changes have happened since my commit,  so I'm not sure i can add much... apart from saying as before I dont think its alsa.

I guess suspicion might be the new midi write thread? which of course in the basic midi thread **should be completely inactive**, perhaps this is continually doing something, which is a bit more expensive in alsa?    but thats a total guess really, as it could be any change in the audio thread handling etc.

**Comment by giuliomoro** [comment]
Fixed 3ed5169.

I was not incrementing the readPointer with the value of bytes written, so what was happening is that the loop was writing every 1ms all the bytes in the buffer. Thanks for your help.

Still, it is interesting that write() was using much less CPU than ALSA, but we do not know what the actual throughput was, so we cannot conclude that ALSA takes more CPU.

**Pull request merged**